File: leetcode/49. Group Anagrams/group-anagrams.py
# 49. Group Anagrams (12/26/56286)
# Runtime: 75 ms (50.91%) Memory: 15.52 MB (76.83%) 

class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        
        anagrams = {}
        for word in strs:
            # Sorted() returns a sorted list of the specified iterable object.
            # It's not hashable, so cast to str()
            sorted_anagram = str(sorted(word))
            if sorted_anagram not in anagrams:
                anagrams[sorted_anagram] = [word]
            else:
                anagrams[sorted_anagram].append(word)

        return anagrams.values()

File: 49. Group Anagrams/group-anagrams.cpp
// 49. Group Anagrams (12/27/56553)
// Runtime: 19 ms (82.05%) Memory: 23.90 MB (67.37%) 

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) 
    {
        std::unordered_map<std::string, std::vector<std::string>> groupedAnagrams;

        for(const auto& str : strs)
        {
            auto sorted = str;
            std::sort(sorted.begin(), sorted.end());

            groupedAnagrams[sorted].push_back(str);
        }

        vector<vector<string>> ret;
        for (const auto& [_, vect] : groupedAnagrams)
        {
            ret.push_back(vect);
        }

        return ret;
    }

};